### ISSUE TITLE
scatter_add requires dimension numbers, default value invalid

### ISSUE THREAD

**Issue by jonasrauber** [google/jax#428]
```python3
jax.lax.scatter_add(np.zeros(10), np.asarray([3, 1]), np.asarray([5., 7.]))
```

fails with `AssertionError` because of `assert type(dimension_numbers) is ScatterDimensionNumbers`

The following works, but there is no documentation on these parameters, so I inferred them from the tests.

```python3
dnums = jax.lax.ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,), index_vector_dim=1)
jax.lax.scatter_add(np.zeros(10), np.asarray([3, 1]), np.asarray([5., 7.]), dnums)
```

Also, it might be worth adding `jax.lax.scatter` with the first argument set to zeros.

**Comment by hawkinsp** [comment]
Thanks for the issue report!

I'll add doc strings for these functions (we haven't gotten around to documenting all of `lax` yet), but essentially gather and scatter_add are direct wrappers around XLA's [gather](https://www.tensorflow.org/xla/operation_semantics#gather) and [scatter](https://www.tensorflow.org/xla/operation_semantics#scatter) primitives, so you can refer to the XLA documentation for details. That said, I personally find their APIs too complicated. Expect us to simplify these APIs in the future.

For most `gather` use cases, it's probably better to use NumPy indexing syntax, which will lower to gather already.

For scatter, unfortunately we don't yet have an equivalent to the gather syntax, but we definitely plan to add some kind of update syntax that looks a lot more like NumPy indexed updates! The closest I can offer you right now is some gradient trickery that exploits the fact that scatter-add is the transpose of gather, and we do have nice indexing support for gather:

```
In [49]: x = np.ones((10,), onp.float32)

In [50]: y = onp.array([2, 4, 5])

In [51]: def scatter_add(f, base, updates):
    ...:     _, grad = jax.vjp(f, base)
    ...:     (out,) = grad(updates)
    ...:     return base + out
    ...:

In [52]: scatter_add(lambda a: a[y], x, np.array([7, 4, 32], np.float32)).copy()
Out[52]: array([ 1.,  1.,  8.,  1.,  5., 33.,  1.,  1.,  1.,  1.], dtype=float32)
```

Expect us to offer something along those lines in the future, but it won't be that exact implementation.

**Comment by hawkinsp** [comment]
PR #429 adds some simple doc strings, although I'm not sure they will clarify much...

We can continue the discussion about more convenient update syntax on
https://github.com/google/jax/issues/101

**Comment by jonasrauber** [comment]
Thanks!
